ibis/expr/operations/udf.py

```diff
@@ -0,0 +1,134 @@
+from typing import TYPE_CHECKING, Any, Callable, TypeVar
+
+from public import public
+
+import ibis.common.exceptions as exc
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.expr.rules as rlz
+from ibis import util
+from ibis.common.annotations import attribute
+from ibis.common.collections import FrozenDict
+
+if TYPE_CHECKING:
+    import ibis.expr.types as ir
+
+
+@public
+class ScalarUDF(ops.Value):
+    @attribute.default
+    def output_shape(self):
+        return rlz.highest_precedence_shape(
+            itertools.chain.from_iterable(
+                util.promote_list(getattr(self, name)) for name in self.argnames
+            )
+        )
+
+
+@public
+class SingleScalarUDF(ScalarUDF):
+    pass
+
+
+@public
+class VectorizedScalarUDF(ScalarUDF):
+    pass
+
+
+def _wrap(wrapper, fn: Callable | None = None, *args: Any, **kwargs: Any) -> Callable:
+    if fn is None:
+        return lambda fn: functools.update_wrapper(wrapper(fn, *args, **kwargs), fn)
+    return functools.update_wrapper(wrapper(fn, *args, **kwargs), fn)
+
+
+S1 = TypeVar("S1", bound=ScalarUDF)
+S2 = TypeVar("S2", bound=ScalarUDF)
+
+
+class UDF(abc.ABC):
+    @util.experimental
+    def __call__(
+        self, fn: Callable | None = None, *args: Any, **kwargs: Any
+    ) -> Callable:
+        return _wrap(self.make_wrapper, fn, *args, **kwargs)
+
+    @util.experimental
+    def vectorized(
+        self, fn: Callable | None = None, *args: Any, **kwargs: Any
+    ) -> Callable:
+        """Construct a vectorized scalar user-defined function."""
+        return _wrap(self.make_vectorized_wrapper, fn, *args, **kwargs)
+
+    @abc.abstractmethod
+    def make_vectorized_wrapper(
+        self, fn: Callable, *args: Any, **kwargs: Any
+    ) -> Callable:
+        ...
+
+    @abc.abstractmethod
+    def make_wrapper(self, fn: Callable, *args: Any, **kwargs: Any) -> Callable:
+        ...
+
+    def make_node(self, fn: Callable, *args, base: type[S1], **kwargs) -> type[S2]:
+        annotations = typing.get_type_hints(fn)
+        if (return_annotation := annotations.pop("return", None)) is None:
+            raise exc.MissingReturnAnnotationError(fn)
+
+        fields = {}
+
+        for name, param in inspect.signature(fn).parameters.items():
+            if (raw_dtype := annotations.get(name)) is None:
+                raise exc.MissingParameterAnnotationError(fn, name)
+
+            arg = rlz.value(dt.dtype(raw_dtype))
+            if (default := param.default) is not EMPTY:
+                arg = rlz.optional(arg, default=default)
+            fields[name] = arg
+
+        fields["output_dtype"] = dt.dtype(return_annotation)
+
+        fields["__func__"] = fn
+        fields["__config__"] = FrozenDict(args=args, kwargs=FrozenDict(**kwargs))
+        fields["__namespace__"] = kwargs.get("schema")
+        fields["__is_server_defined__"] = kwargs.get("__is_server_defined__", False)
+
+        return type(fn.__name__, (base,), fields)
```

Comment:
IIUC this is defining a new subclass for every UDF. I'm a bit worried about this approach for the following reasons:

- Dynamically defined types won't serialize (easily), preventing the operation tree from being pickled using `pickle`. This may not be important now, but if it was in the future then this approach would be a blocker.
- It's a bit complex. It took me a few readings of this implementation to understand all the plumbing here.

Would it be possible to avoid creating dynamic subclasses, and instead plumb through the function info as attributes on *instances* of the various operations, instead of storing them as attributes on dynamically generated classes? This would mean that all scalar vectorized UDF calls would result in instances of `VectorizedScalarUDF`, rather than instances of custom subclasses for each `UDF` definition.

Reply:
The primary issue with this is that there isn't way to hook into the validation system without creating fields on a class.

All of the machinery to validate inputs is created based on a class definition and can't really be moved to runtime without weird hacks AFAICT.

Reply:
So, while I can plumb almost everything through as instance attributes, the signature is unique to an operation class.

Reply:
The other thing to note is that validation happens during operation construction, so constructing an instance that doesn't validate anything (because the signature would have to live elsewhere) would just bypass all validation.